test/leak.cc

```diff
@@ -0,0 +1,181 @@
+
+#include <memory>
+#include <string>
+
+#include "include/proxy-wasm/context.h"
+#include "include/proxy-wasm/wasm.h"
+#include "include/proxy-wasm/wasmtime.h"
+
+#include "wasmtime/include/wasm.h"
+#include "wasmtime/include/wasmtime.h"
+
+#include "test/watdata.h"
+#include "gtest/gtest.h"
+#include <chrono>
+#include <fstream>
+#include <thread>
+
+namespace proxy_wasm {
+namespace {
+
+struct DummyIntegration : public WasmVmIntegration {
+  ~DummyIntegration() override{};
+  WasmVmIntegration *clone() override { return new DummyIntegration{}; }
+  void error(std::string_view message) override {
+    std::cout << "ERROR from integration: " << message << std::endl;
+    error_message_ = message;
+  }
+  bool getNullVmFunction(std::string_view function_name, bool returns_word, int number_of_arguments,
+                         NullPlugin *plugin, void *ptr_to_function_return) override {
+    return false;
+  };
+  std::string error_message_;
+};
+
+void callback(void *raw_context) {}
+Word callback2(void *raw_context, Word val) { return val + 100; }
+
+TEST(Wasmtime, CreateVM) {
+  auto raw = wat_data::HelloGo;
+
+  wasm_byte_vec_t wat;
+  wasm_byte_vec_new_uninitialized(&wat, raw.size());
+  ::memcpy(wat.data, raw.data(), raw.size());
+
+  wasm_byte_vec_t source;
+  wasmtime_error_t *error = wasmtime_wat2wasm(&wat, &source);
+  EXPECT_TRUE(error == nullptr);
+  auto binary = std::string(source.data, source.size);
+  wasm_byte_vec_delete(&wat);
+  wasm_byte_vec_delete(&source);
+
+  WasmCallVoid<0> run;
+  WasmCallWord<1> run2;
+
+  for (int i = 0; i < 1000; i++) {
```

Comment:
I cannot observe any memory leak with this loop, so the vm/module setup in Envoy does not leak anything. I am now suspicious about memory operations

Reply:
or function calls